### DISCUSSION TITLE
run_sync seems to progressively slow down application, does it block the event loop until server does not reply?

### DISCUSSION THREAD

**Discussion by pankajkumar229** [sqlalchemy/sqlalchemy#9285] [Usage Questions] [answered]
We are using async_scoped_session to talk to Postgres in a tornado web application. In one of the previous discussions, we found that we can use AsyncSession's run_sync method to load extra arguments as needed: https://github.com/sqlalchemy/sqlalchemy/discussions/9166 .

It was discussed that run_sync will not block the event loop. The problem I see that if we call run_sync method to load more ORM objects in a tornado request handler, the response time(latency) of the application becomes progressively slower. Could run_sync be blocking the tornado event loop for the time SQL response is not received? It is expected that latency increases for a synchronous blocking version of tornado. It seems to be giving similar behavior when using run_sync.

**Comment by pankajkumar229** [comment]
Does run_sync run on a thread pool or something that limits the parallelism somehow?

**Comment by CaselIT** [comment, marked as answer]
Hi,

If used with sqlalchemy operations run_sync does not block the asyncio loop. It does the exact same thing as when calling any other sqlalchemy async operation.

You cannot use other blocking calls in a run_sync function though (no sleep or blocking network calls).

**Reply by pankajkumar229** [reply]
All I am using run_sync for is to reload session objects. I can show you that as soon I start calling this function, the latency starts increasing forever. 

I am suspecting run_async because:
- There is another await session.execute in the request handler but that does not cause this behavior. SQL seems to be super fast at other places when querying.
- Latency keeps increasing when a fixed number of parallel clients are requesting. 

I will check more

**Reply by CaselIT** [reply]
I'm not sure what "latency starts increasing forever." mean in this context. Consider that a refresh operation may run sql queries, so an increase in time is normal.

Without an example of use it's hard to provide any suggestions.

**Reply by pankajkumar229** [reply]
I was discussing with a friend and we had a doubt. We are writing a function which we run with run_sync . This function uses getattr to access various attributes that we need to load like in the sync version of sqlalchemy. But every invocation of getattr possibly does a SQL call. 

In regular async, we should await in between all those SQL calls so the event loop continues. Here we just do it one after the other. Would the event loop not be blocked? I will put a more detailed view of what we are doing if the refresh method does not work.

**Reply by CaselIT** [reply]
no, each execute is proceeding like it would when calling await. so the loop is not blocked

**Reply by pankajkumar229** [reply]
I will try to put the description code as briefly as I can. Please let me know if I missed something or have an error.

Problem:
We have a tornado + sqlalchemy app which we are accessing with an async_scoped_session. We are putting load using 200 threads on a client. If we don't do any SQLAlchemy calls, the throughput is about 100 calls/second with a 50-100 ms response time. Then we introduce a single execute call to load one row from the user table and populate a user object. The throughput goes down and response time increases but that is ok. The real problem occurs when we try to load many attributes using run_sync. The response time on the server keeps increasing to a very large number(100 secs ) . See below that the latency is doubling.
```
#######%%%%%% 100 calls took 161.223754 seconds throughput of 0.620255995279703calls/second
#######%%%%%% Average Latency 36.66651808000002
#######%%%%%%
#######%%%%%% 100 calls took 62.057452 seconds throughput of 1.611410020508093calls/second
#######%%%%%% Average Latency 37.2427997
#######%%%%%%
#######%%%%%% 100 calls took 53.237594 seconds throughput of 1.8783718888573364calls/second
#######%%%%%% Average Latency 38.118982890000005
#######%%%%%%
#######%%%%%% 100 calls took 70.79139 seconds throughput of 1.412601165198197calls/second
#######%%%%%% Average Latency 42.70391274999999
#######%%%%%%
#######%%%%%% 100 calls took 64.951872 seconds throughput of 1.5396015067895197calls/second
#######%%%%%% Average Latency 46.81780110999999
#######%%%%%%
#######%%%%%% 100 calls took 75.560282 seconds throughput of 1.323446622393495calls/second
#######%%%%%% Average Latency 44.08126153999997
#######%%%%%%
#######%%%%%% 100 calls took 97.382516 seconds throughput of 1.026878377223279calls/second
#######%%%%%% Average Latency 43.190470899999994
#######%%%%%%
#######%%%%%% 100 calls took 97.418576 seconds throughput of 1.0264982727729464calls/second
#######%%%%%% Average Latency 56.56797211999998
#######%%%%%%
#######%%%%%% 100 calls took 154.012819 seconds throughput of 0.6492966017328726calls/second
#######%%%%%% Average Latency 57.55805273000001
#######%%%%%%
#######%%%%%% 100 calls took 76.905211 seconds throughput of 1.3003020042425995calls/second
#######%%%%%% Average Latency 53.86179309
#######%%%%%%
#######%%%%%% 100 calls took 114.177966 seconds throughput of 0.8758257263051963calls/second
#######%%%%%% Average Latency 57.500908549999984
#######%%%%%%
#######%%%%%% 100 calls took 128.466246 seconds throughput of 0.7784145883736651calls/second
#######%%%%%% Average Latency 62.54195279000001
```

Details with pseudocode where it happens:
```
## ORM class definitions
class User
     courses = ## relationship to Course

class Course
...

### Book class does all db calls and maintains async_scoped_session
class Book:
    @classmethod
    engine = create_async_engine(db_url, pool_size=400, pool_recycle=3600)
    async_session_factory = sessionmaker(autoflush=True, bind=self.engine, class_=AsyncSession)
    bookconnection = async_scoped_session( async_session_factory, scope = current_task)


    async def find_user():
             bookconnection.execute(select(User))
             ...
             return user

     async def load_attributes_to_object2(self, main_obj, attrs):
         next_objs = [main_obj]
         #print("To load " + str(attrs))
         def load_attributes_sync(session) :
             nonlocal next_objs, attrs
             for x in attrs:
                 print(" Processing attribute " + str(x))
                 new_next_objs = []
                 if isinstance(x, str):
                     for next_obj in next_objs:
                         new_next_obj = getattr(next_obj, x)
                         new_next_objs.append(new_next_obj)
                 elif isinstance(x, list):
                     x_attr_name = x[0]
                     for next_obj in next_objs:
                         list_new_next_obj = getattr(next_obj, x_attr_name)
                         for new_next_obj in list_new_next_obj:
                             new_next_objs.append(new_next_obj)
                 #print("After processing " + str(x) + ", New next objs length = " + str(len(new_next_objs)))
                 next_objs = new_next_objs

         #await asyncio.to_thread(load_attributes_sync)
         await self.bookconnection().run_sync(load_attributes_sync)

### BaseHandler maintains a static handle to Book class
class BaseHandler:
      @classmethod
      mybook = Book()

### Tornado response handler when someone requests user details.
class UserHandler(tornado.RequestHandler, BaseHandler)
    async def get():
         user = BaseHandler.mybook.find_user(req.query.email)  ### Line 1
         BaseHandler.load_attributes_to_object(user, ["courses" , "assignments", "courses_taught" ...]) ### Line 2
         return success ### Line 3

```

Now , we run a client with 200 threads putting load on this server calling user get over and over with a sleep of 3 seconds after every request. If we comment both "Line 1" and "Line 2" in "async def get()", we have a throughput of 100 req/sec with ~50-100ms response time. If we bring "Line 1" back, the response time increases and throughput decreases but it keeps constant. IF we uncomment all lines in the function "async def get()" , the response time starts increasing and throughput comes to less than 1 req/sec as seen in the problem definition. The database itself is not overloaded. The memory/cpu on the python process is also stable.

I am little stumped why response time keeps increasing.

**Reply by pankajkumar229** [reply]
Also, after a while we get the error : "sqlalchemy.exc.TimeoutError: QueuePool limit of size 400 overflow 10 reached, connection timed out, timeout 30.00 (Background on this error at: https://sqlalche.me/e/14/3o7r)"

There are 200 parallel clients. Should not a pool size of 400 be enough?

**Reply by pankajkumar229** [reply]
![Screen Shot 2023-02-12 at 6 14 37 PM](https://user-images.githubusercontent.com/1482916/218355520-74a78362-06b8-4f61-8cb7-fb1bba9ba014.png)

My apologies for too many messages but hoping some angels here might just be able to see whats going on. I also see that the number of async tasks keeps increasing. Not sure how that can happen when we have a limit of 200 parallel requests. It should spawn 200 tasks which should get over when the request finishes?

**Reply by pankajkumar229** [reply]
Could it be that run_sync is causing some leaking tasks? Could it be that calling run_sync causes the executing task to never get over?

I did some analysis of the tasks in bad and good case and it seems to be the suspicion. 
```
##### Out of all the async tasks, here is the breakup when there is lot of delay.

(Pdb) len(asyncio.all_tasks())
529
(Pdb) len({str(x.get_stack()) for x in asyncio.all_tasks() if "1713" in str(x.get_stack())})
200 – opened by tornado (good)
(Pdb) len({str(x.get_stack()) for x in asyncio.all_tasks() if "825" in str(x.get_stack())})
200 – opened by tornado (good)
(Pdb) len({str(x.get_stack()) for x in asyncio.all_tasks() if "700" in str(x.get_stack())})
7 (small enough, opened by asyncpg)
(Pdb) len({str(x.get_stack()) for x in asyncio.all_tasks() if "817" in str(x.get_stack())})
128
(Pdb) len({str(x.get_stack()) for x in asyncio.all_tasks() if "labgrader_appserver" in str(x.get_stack()) })
2


(Pdb)  len(asyncio.all_tasks())
880
(Pdb) len({str(x.get_stack()) for x in asyncio.all_tasks() if "1713" in str(x.get_stack())})
542
(Pdb) len({str(x.get_stack()) for x in asyncio.all_tasks() if "825" in str(x.get_stack())})
141
(Pdb) len({str(x.get_stack()) for x in asyncio.all_tasks() if "817" in str(x.get_stack())})
2
(Pdb)


#### In a good case when we dont call run_sync, it looks like this
(Pdb) import asyncio
(Pdb) len(asyncio.all_tasks())
39
(Pdb)  len({str(x.get_stack()) for x in asyncio.all_tasks() if "1713" in str(x.get_stack())})
0
(Pdb)  len({str(x.get_stack()) for x in asyncio.all_tasks() if "825" in str(x.get_stack())})
38
(Pdb) len({str(x.get_stack()) for x in asyncio.all_tasks() if "817" in str(x.get_stack())})
0

```

**Reply by pankajkumar229** [reply]
PLease note in code the extra () around async_scoped_session used to call run_sync : await self.bookconnection().run_sync(load_attributes_sync) . I believe you suggested before we could access the one from AsyncSession since run_sync was not available on async_scoped_session directly. 

Could this be causing the leaking tasks?

**Reply by zzzeek** [reply]
it seems clear your program is not returning connections to the pool.    People report this kind of problem quite often when using `async_scoped_session` because their program does not correctly handle the context and/or the lifespan of each session within the scope.  

I would ensure that a single AsyncSession is created at the start of a request, linked with the request, then definitely torn down at the end of a request.  I'd try to make this work without using async_scoped_session, which is attempting to create a "global variable" that's generally an anti-pattern.

**Reply by pankajkumar229** [reply]
I will try your suggestion and report. Thanks

**Reply by zzzeek** [reply]
you also want to be looking at [pg_stat_activity](https://www.postgresql.org/docs/current/monitoring-stats.html#MONITORING-PG-STAT-ACTIVITY-VIEW) while the program is running.  that will show you exactly what's going on, if queries are hanging / deadlocking or if many connections are "idle in transaction", which would indicate connections not being returned.

as far as "run_sync" itself, all database calls in SQLAlchemy run within "run_sync" when using the asyncio API.    see the box "asyncio and events, two opposites" at the bottom of https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#examples-of-event-listeners-with-async-engines-sessions-sessionmakers .

**Comment by pankajkumar229** [comment]
Hi Mike and Federico,

Thanks so much for all your help and I am really sorry for all the doubts on the framework. I found the bug in our usage of run_sync. We were somehow using append on a list which we thought was a local variable. But it was actually a class member variable. This caused repeats on the number of arguments. The first one was ('courses', 'courses_taught'). In the second call, the list of argument became ('courses', 'courses_taught', 'courses', 'courses_taught') and kept increasing. This caused run_sync loop to go forever. We have fixed it since and I reverted back to async_scoped_session, things work well.

We have been doing a large refactor and sometimes running into bugs that are hard to pinpoint.


Many Thanks,
